atl-checker/src/algorithms/game_strategy/partial.rs

```diff
@@ -0,0 +1,275 @@
+use crate::algorithms::certain_zero::common::VertexAssignment;
+use crate::algorithms::game_strategy::format::PartialStrategyWithFormatting;
+use crate::atl::Phi;
+use crate::edg::annotated_edg::{AnnotatedEdge, AnnotatedExtendedDependencyGraph};
+use crate::edg::atledg::pmoves::PartialMove;
+use crate::edg::atledg::vertex::AtlVertex;
+use crate::edg::atledg::AtlDependencyGraph;
+use crate::game_structure::{GameStructure, Player, State};
+use std::collections::HashMap;
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct PartialStrategy {
+    /// The players for which this strategy applies
+    pub players: Vec<Player>,
+    /// A partial mapping from States to a partial move, where the given players have made
+    /// a specific choice.
+    pub move_to_pick: HashMap<State, PartialMove>,
+}
+
+impl PartialStrategy {
+    /// Pairs a partial strategy with its game structure, allowing us to print
+    /// the strategy using the names of players and moves as defined by the game structure.
+    pub fn in_context_of<'a, G: GameStructure>(
+        &'a self,
+        game: &'a G,
+    ) -> PartialStrategyWithFormatting<'a, G> {
+        PartialStrategyWithFormatting {
+            strategy: self,
+            game,
+        }
+    }
+}
+
+/// Computes the PartialStrategy that the given players can use to enforce or break
+/// the given property. In other words, the behaviour of this function is undefined, if the
+/// ATL formula of v0 is an unsatisfied enforce property or a satisfied despite property.
+pub fn compute_partial_strategy<G: GameStructure>(
+    graph: &AtlDependencyGraph<G>,
+    v0: &AtlVertex,
+    assignments: &HashMap<AtlVertex, VertexAssignment>,
+) -> PartialStrategy {
+    let mut move_to_pick = HashMap::new();
+
+    compute_partial_strategy_rec(graph, v0, assignments, &mut move_to_pick);
+
+    PartialStrategy {
+        players: v0.formula().players().unwrap().into(),
+        move_to_pick,
+    }
+}
+
+/// Recursive helper function to [compute_partial_strategy].
+fn compute_partial_strategy_rec<G: GameStructure>(
+    graph: &AtlDependencyGraph<G>,
+    vertex: &AtlVertex,
+    assignments: &HashMap<AtlVertex, VertexAssignment>,
+    move_to_pick: &mut HashMap<State, PartialMove>,
+) {
+    match vertex {
+        AtlVertex::Full { state, formula } => {
+            if move_to_pick.get(state).is_some() {
+                // We have already found the move to pick in this state
+                return;
+            }
+
+            match formula.as_ref() {
+                Phi::EnforceNext { .. } => {
+                    let edges = graph.annotated_succ(vertex);
+                    // Find the first hyper-edge where all targets are true
+                    for edge in edges {
```

Comment:
This I have no opinion on, but it could be: (and save us an assignment)
```suggestion
                    // Find the first hyper-edge where all targets are true
                    for edge in graph.annotated_succ(vertex){
```